### ISSUE TITLE
Can't login with LDAP users

### ISSUE THREAD

**Issue by melo0187** [gitblit/gitblit#1013]
I am struggling to log in to GitBlit with my LDAP users.
I run Gitblit WAR 1.7.1 in Tomcat 7.0.62 on my Synology DiskStation running DSM 5.2 and did specify baseFolder via JNDI in Tomcat's context.xml, which works fine. I was able to browse to GitBlit, log in with admin/admin and changed the password.

I went ahead and did edit the gitblit.properties file located in the path I specified as baseFolder via JNDI. I did add overrides to the settings required for LDAP to connect to the Directory Server running on the DiskStation:
```
realm.userService = com.gitblit.LdapUserService
realm.authenticationProviders = ldap
realm.ldap.server = ldap://localhost
realm.ldap.username = uid=root,cn=users,dc=melo,dc=myds,dc=me
realm.ldap.password = redacted
realm.ldap.maintainTeams = false
realm.ldap.accountBase = dc=melo,dc=myds,dc=me
realm.ldap.admins = admin
```
I did choose these values based on what information I have from DSM Directory Server UI:
```
Base DN: dc=melo,dc=myds,dc=me
Bind DN: uid=root,cn=users,dc=melo,dc=myds,dc=me
```

After a restart GitBlit comes back up, but I can log in only with the admin user in user.conf. When I try to log in with one of my Directory Server users (admin or melo) I get "Invalid username or password!".

I thought there might be a collision because I have admin as a username both in user.conf and in the Directory Server, so I changed the username in user.conf and restarted GitBlit. The change took effect and I can only use the new username to log in, but logging in with my Directory Server users still fails.

I then tried to set `realm.ldap.synchronize = true` and restarted GitBlit. I saw that user.conf had new entries corresponding to my Directory Server users. However GitBlit did not manage to finish startup. I had to revert to `realm.ldap.synchronize = false` and clean the user.conf file from the external users, before GitBlit successfully started again.
Because of the sync adding entries to my user.conf file I believe that my realm.ldap settings might be correct. However, I can't use the Directory Server users to log in to GitBlit.

Can anyone please assist me with this? Is there something obviously wrong with my settings? Did I misunderstand how this is supposed to work?
I actually don't even need to have local users in users.conf and would be fine to only use the users specified in my Directory Server.

Thanks in advance for any help.

**Comment by gitblit** [comment]
Did you set this?

realm.authenticationProviders=ldap

Have you reviewed the logs to see if there are any reported errors?

-J

**Comment by melo0187** [comment]
@gitblit thanks for your reply.

I did set `realm.authenticationProviders=ldap` (as described in the first post).

However I am not sure where to check for logs. If you can tell me where I can find gitblit logs (I run the GitBlit WAR on Tomcat 7, so I am not really sure which log to check), I would be happy to review them the log an report back.

**Comment by gitblit** [comment]
Please double-check `realm.authenticationProviders`.

In your first post you specified `realm.userService = com.gitblit.LdapUserService` which is incorrect.  It should be `realm.authenticationProviders=ldap`.

The WAR variant will log to the container's log.  In the case of Tomcat I think it is *catalina.out* but there is great variability with that so I can't tell you exactly where your Tomcat is logging - especially on a Synology NAS.

**Comment by melo0187** [comment]
@gitblit in my first post I specified both `realm.authenticationProviders=ldap` and `realm.userService = com.gitblit.LdapUserService`. 
Do you mean I need to remove realm.userService and just use realm.authenticationProviders?

I can check the catalina.out log. I know where that one is on my NAS. Will report back.
Thanks for your help.

**Comment by gitblit** [comment]
Ah. So you did; I missed that.

You'll want to restore `realm.userService=${baseFolder}/users.conf`.  You are mixing & matching old configuration with current configuration.  Did you find that info on the site somewhere or are you migrating an old setup to a new one?

**Comment by melo0187** [comment]
I did remove `realm.userService` from the file all together, since it is set in default.properties.

You are right that I am mixing up old and current config. I just noticed when looking up the site again:
http://gitblit.com/setup.html

This seems to be from version 1.2.1. Sadly it is the second hit on google when searching for _gitblit ldap not working_ and I did not spot the old version number on top.

Anyway, I removed `realm.userService` from gitblit.properties and restarted gitblit, but the problem still persists. I just grabbed catalina.out but it is late. I will review it tomorrow. 
Just to make sure I uploaded my gitblit.properties file. [https://gist.github.com/melo0187/9c5aaa8a1ba593ed7571](url)
Reviews are welcome and thanks again for your help.

**Comment by melo0187** [comment]
Still struggling. I redeployed GitBlit with the config posted in the last comment.
Gitblit comes up, log looks normal and says things like
```
[INFO ] ----[com.gitblit.manager.IAuthenticationManager]----
[INFO ] setting up com.gitblit.auth.LdapAuthProvider
[INFO ] Ldap sync service is disabled.
```
When I try to log in, this is logged:
`[WARN ] Failed login attempt for melo, invalid credentials from 78.43.225.169`

Why? I have no idea.
I uploaded the log [here](https://gist.github.com/melo0187/55df027594c09d2c74cd).
Maybe you can have a look at it?
Any feedback is much appreciated.

**Comment by melo0187** [comment]
Finally I managed to get this done!

It was a small change, but one I don't really understand. I am not really proficient at using LDAP.
I just added this to _gitblit.properties:_
`realm.ldap.accountPattern = (&(objectclass=inetorgperson)(cn=${username}))`

This overrides the default from _default.properties_, which looks like this:
`realm.ldap.accountPattern = (&(objectClass=person)(sAMAccountName=${username}))`

So apparently the default query is not applicable for Synology's Directory Server.
Any idea how I could have figured this one out myself? I simply copy-pasted that accountPattern from a post in google groups... 

Anyway, it is working now and I am very excited to start working with GitBlit. It looks really promising =)

Thanks for your time and assistance!
